internal/errgrouputil/builder.go

```diff
@@ -0,0 +1,54 @@
+// Package errgrouputil contains methods for working with errgroup code.
+package errgrouputil
+
+import (
+	"context"
+	"runtime"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/pomerium/pomerium/pkg/slices"
+)
+
+// BuilderFunc is a function that builds a value of type T
+type BuilderFunc[T any] func(ctx context.Context) (*T, error)
+
+// Build builds a slice of values of type T using the provided builders concurrently
+// and returns the results and any errors.
+func Build[T any](
+	ctx context.Context,
+	builders ...BuilderFunc[T],
+) ([]*T, []error) {
+	eg, ctx := errgroup.WithContext(ctx)
+	eg.SetLimit(runtime.NumCPU()/2 + 1)
+
+	results := make([]*T, len(builders))
+	errors := make([]error, len(builders))
+
+	fn := func(i int) func() error {
+		return func() error {
+			result, err := builders[i](ctx)
```

Comment:
I'm not sure about this context propagation either. Would it be better to check for cancellation here (or around line 41?), and return early if cancelled, without the builder function needing to know about the errgroup context?